File: domain.guice/src/main/java/org/openmedstack/domain/guice/HttpEventStore.java
package org.openmedstack.domain.guice;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.openmedstack.eventstore.*;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class HttpEventStore implements IStoreEvents, ICommitEvents {
    private final HttpClient _persistence;
    private final ObjectMapper _mapper;
    private final URI _baseAddress;

    public HttpEventStore(ObjectMapper mapper, URI baseAddress) {
        _baseAddress = baseAddress;
        _persistence = HttpClient.newHttpClient();
        _mapper = mapper;
    }

    @Override
    public CompletableFuture<IEventStream> createStream(String bucketId, String streamId) {
        return CompletableFuture.completedFuture(OptimisticEventStream.Create(bucketId, streamId, this));
    }

    @Override
    public CompletableFuture<IEventStream> openStream(String bucketId, String streamId, int minRevision, int maxRevision) {
        if (streamId == null) {
            throw new NullPointerException();
        }
        maxRevision = maxRevision <= 0 ? Integer.MAX_VALUE : maxRevision;

        return OptimisticEventStream.Create(bucketId, streamId, this, minRevision, maxRevision);
    }

    @Override
    public CompletableFuture<IEventStream> openStream(Snapshot snapshot, int maxRevision) {
        return openStream(snapshot.getBucketId(), snapshot.getStreamId(), 0, maxRevision <= 0 ? Integer.MAX_VALUE : maxRevision);
    }

    @Override
    public void close() throws Exception {
    }

    @Override
    public CompletableFuture<Iterable<Commit>> getFrom(String bucketId, String streamId, int minRevision, int maxRevision) {
        var request = HttpRequest.newBuilder(
                        URI.create(String.format("/%s/%s/%d/%d", bucketId, streamId, minRevision, maxRevision))
                                .relativize(_baseAddress))
                .GET()
                .build();
        return _persistence.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApplyAsync(HttpResponse::body)
                .thenApplyAsync(c -> {
                    try {
                        return _mapper.readValue(c, Commit[].class);
                    } catch (JsonProcessingException e) {
                    }
                    return new Commit[0];
                }).thenApplyAsync(c -> List.of(c));
    }

    @Override
    public CompletableFuture<Commit> commit(CommitAttempt attempt) {
        try{
            var json = _mapper.writeValueAsString(attempt);
        var request = HttpRequest.newBuilder(_baseAddress).POST(HttpRequest.BodyPublishers.ofString(json)).build();
        return _persistence.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApplyAsync(HttpResponse::body)
                .thenApplyAsync(c -> {
                    try {
                        return _mapper.readValue(c, Commit.class);
                    } catch (JsonProcessingException e) {
                        e.printStackTrace();
                    }
                    return null;
                });}
        catch (JsonProcessingException j){
            return CompletableFuture.completedFuture(null);
        }
    }
}

File: domain.guice/src/main/kotlin/org/openmedstack/domain/guice/HttpEventStore.kt
package org.openmedstack.domain.guice

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.inject.Inject
import org.openmedstack.domain.Memento
import org.openmedstack.eventstore.*
import java.net.*
import java.net.http.*
import java.util.concurrent.CompletableFuture

class HttpEventStore @Inject constructor(mapper: ObjectMapper) : IStoreEvents, ICommitEvents {
    private val _persistence: HttpClient
    private val _mapper: ObjectMapper
    override fun createStream(bucketId: String?, streamId: String): CompletableFuture<IEventStream> {
        return CompletableFuture.completedFuture(OptimisticEventStream.create(bucketId, streamId, this))
    }

    override fun openStream(bucketId: String?, streamId: String, minRevision: Int, maxRevision: Int): CompletableFuture<IEventStream> {
        val mr = if (maxRevision <= 0) Int.MAX_VALUE else maxRevision
        return OptimisticEventStream.create(bucketId, streamId, this, minRevision.coerceAtLeast(0), mr)
    }

    override fun <TMemento : Memento> openStream(snapshot: Snapshot<TMemento>, maxRevision: Int): CompletableFuture<IEventStream> {
        return openStream(snapshot.bucketId, snapshot.streamId, 0, if (maxRevision <= 0) Int.MAX_VALUE else maxRevision)
    }

    @Throws(Exception::class)
    override fun close() {
    }

    override fun getFrom(bucketId: String?, streamId: String, minRevision: Int, maxRevision: Int): CompletableFuture<Iterable<Commit>> {
        val request = HttpRequest.newBuilder(URI.create("")).GET().build()
        return _persistence.sendAsync(request, HttpResponse.BodyHandlers.ofString())
            .thenApplyAsync { obj: HttpResponse<String> -> obj.body() }
            .thenApplyAsync { c: String -> _mapper.readValue(c, Array<Commit>::class.java).asList() }
    }

    override fun commit(attempt: CommitAttempt): CompletableFuture<Commit> {
        return try {
            val json = _mapper.writeValueAsString(attempt)
            val request = HttpRequest.newBuilder(URI.create("")).POST(HttpRequest.BodyPublishers.ofString(json)).build()
            _persistence.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                    .thenApplyAsync { obj: HttpResponse<String?> -> obj.body() }
                    .thenApplyAsync { c: String? ->
                        try {
                            return@thenApplyAsync _mapper.readValue(c, Commit::class.java)
                        } catch (e: JsonProcessingException) {
                            e.printStackTrace()
                        }
                        null
                    }
        } catch (j: JsonProcessingException) {
            CompletableFuture.completedFuture(null)
        }
    }

    init {
        _mapper = mapper
        _persistence = HttpClient.newHttpClient()
    }
}